homeassistant/components/heos/media_player.py

```diff
@@ -1,152 +1,262 @@
+        """Mute the volume."""
+        await self._player.set_mute(mute)
+
+    async def async_set_shuffle(self, shuffle):
+        """Enable/disable shuffle mode."""
+        await self._player.set_play_mode(self._player.repeat, shuffle)
 
-    def _update_state(self):
-        self.async_schedule_update_ha_state()
+    async def async_set_volume_level(self, volume):
+        """Set volume level, range 0..1."""
+        await self._player.set_volume(volume * 100)
 
     async def async_update(self):
-        """Update the player."""
-        self._player.request_update()
+        """Update supported features of the player."""
+        from pyheos import const
+        control_to_support = {
+            const.CONTROL_PLAY: SUPPORT_PLAY,
+            const.CONTROL_PAUSE: SUPPORT_PAUSE,
+            const.CONTROL_STOP: SUPPORT_STOP,
+            const.CONTROL_PLAY_PREVIOUS: SUPPORT_PREVIOUS_TRACK,
+            const.CONTROL_PLAY_NEXT: SUPPORT_NEXT_TRACK
+        }
+        controls = self._player.now_playing_media.supported_controls
+        current_support = [control_to_support[control] for control in controls]
+        self._supported_features = reduce(ior, current_support,
+                                          BASE_SUPPORTED_FEATURES)
+
+    async def async_will_remove_from_hass(self):
+        """Disconnect the device when removed."""
+        for signal_remove in self._signals:
+            signal_remove()
+        self._signals.clear()
 
-    async def async_added_to_hass(self):
-        """Device added to hass."""
-        self._player.state_change_callback = self._update_state
+    @property
+    def available(self) -> bool:
+        """Return True if the device is available."""
+        return self._player.available
 
     @property
-    def unique_id(self):
-        """Get unique id of the player."""
-        return self._player.player_id
+    def device_info(self) -> dict:
+        """Get attributes about the device."""
+        return {
+            'identifiers': {
+                (DOMAIN, self._player.player_id)
+            },
+            'name': self._player.name,
+            'model': self._player.model,
+            'manufacturer': 'HEOS',
+            'sw_version': self._player.version
+        }
 
     @property
-    def name(self):
-        """Return the name of the device."""
-        return self._player.name
+    def device_state_attributes(self) -> dict:
+        """Get additional attribute about the state."""
+        return {
+            'media_album_id': self._player.now_playing_media.album_id,
+            'media_queue_id': self._player.now_playing_media.queue_id,
+            'media_source_id': self._player.now_playing_media.source_id,
+            'media_station': self._player.now_playing_media.station,
+            'media_type': self._player.now_playing_media.type,
+            'player_ip_address': self._player.ip_address,
+            'player_network_connection_type': self._player.network
```

Comment:
Will this change during entity lifetime? We should reserve the state attributes for state related (dynamic) things.

Reply:
It's unlikely, but possible (same with `ip_address`).  Both of those are probably better for `device_info` except it ignores custom keys today.  I can remove for now if that's preferred.

Reply:
Yeah, if they are mostly static and aren't needed for automations they don't fit state attributes, I think.

Reply:
Agreed and removed.